Actions from state

jeremiahboswell opened the pull request:
After the discovery of the _play_turn()_ function in _utils.py_ from class Monday, it was trivial to create a function that would theoretically take some indicator of state and return a list of actions (a list of lists in this case). It isn't much honestly, but I figure especially with keeping the comments regarding my expanding knowledge of the Fireplace API as well as Python, pulling this into the main repository should allow @oostin623 to modify it into something that can be tested through an actual game simulation.

The function takes a player as an argument, which may be problematic if this can not be extracted from the game state's representation. Additionally, with taking a player as an argument and getting all the needed attributes from that object, it did not appear that any imports were needed. There are comments in the actual code section that outline which parts are less likely to pass arguments to the sublists that will be actually useful.

I suspect that this pull request cannot be automatically merged due to @oostin623 's modifications to the file involved removing lines included in the older version I forked despite updating it. I also still do not appear to have write access to the repository, which isn't really a problem with so little time remaining, other than if minor changes need to be made (perhaps moving the research comments into a separate plaintext document?).

Commit:
Added file of relevant research and declarations

Created for more visible access. Writing style of comments made clearer,
more direct. Code does not yet reflect extraction of valid actions from
state. Will be updated in future commits.

Will not currently run due to lack of imports from the API.

Commit:
Updated README.md with new file description

Commit:
Created state_actions function

Function theoretically works based on Fireplace API attributes. Returns
a list containing a list for each different type of action (hero power,
 cards, characters). Comments of previous API research moved to the
bottom of the file.

dillondaudert commented on src/game-interface/state-actions.py:
```diff
@@ -0,0 +1,143 @@
+      if card.must_choose_one:
+        cardactions = card.choose_cards
+      if card.has_target():
+        cardactions = cardactions + card.targets
```
Does card.has_target() cover summoning monsters and playing spells that may not have an explicit target?

jeremiahboswell replied:
I think this is actually covered by the top level check using _is_playable()_. For the latter part of your questions (spells with no target), this is covered by the following conditional within _is_playable()_:

if _PlayReq.REQ_TARGET_TO_PLAY_ in _self.requirements_

The purpose of the _has_target()_ check in this function is to get the different possibilities for what target a card can be used on. A likely example is that the bot can play a certain spell card (they meet the requirements on the card), but it could be used on multiple targets on the field, so all of the different target possibilities are stored in the cardactions list.

dillondaudert commented on src/game-interface/state-actions.py:
```diff
@@ -0,0 +1,143 @@
+  # Characters are defined are involved in functions and properties in player.py
+  for character in player.characters:
+    if character.can_attack():
+      attackactions = character.targets
```
Does this get overwritten every time it goes through the loop? Seems we should append the list of attackactions for each character on the field.

jeremiahboswell replied:
In its current state, this is true. I was focused on looking at one card's possible actions and only seemed to remember that the later checks were inside a loop for each card in the bot's hand for only one check. I will make this change and push it to my branch in order to update this pull request.

Commit:
Appended looped attributes to corresponding lists

Additionally, declared lists present in loops at the beginning of the
function to prevent "not defined" errors.

ghost commented:
I apologize for not responding to your questions earlier. I set up my Github account to use the email I should be checking more, and I must have missed some other notifications. I was fortune enough to check Github today and notice that there were comments added to this pull request. I have pushed another commit to the branch that appends the actions that are present in the two lower loops to their respective lists. I also declared said lists beforehand, noting that failing to do so previously causes a "not defined" error, which I did not expect for Python due to being dynamically typed (I did see this possibility for the last portion where the _append()_ function is used; I am not sure why I did not make similar changes above earlier)

ghost commented:
Closing this pull request due to the likelyhood of the parallel branch already having this functionality.
